Fix#103

SRFU-NN opened the pull request:
WIP

Needs more detailing, SRFU will do this before 2022-12-09

Closes #103

sqbl committed:
Added alternative dependence funcction. Currently shoehorned to fix input architecture from old function

sqbl committed:
Commented on arguments for new dependence function. Added logic around return std

SRFU-NN committed:
chore: Expanded gitignore

SRFU-NN committed:
doc: Argument for sampling strategy added

SRFU-NN committed:
refactor: evenly_sample() to method in dimension

SRFU-NN committed:
chore: Corrected help text in pytest.ini

SRFU-NN committed:
feat: Resintated old dependence()

SRFU-NN committed:
refactor: Leaner code for finding plotting optimum

The code for finding the place for the red dots and red dotted lines
now has fewer repetitions.

SRFU-NN committed:
feat: Different way to make random data

Instead of doing it once per pixel per graph,
do it once for the entire space.

SRFU-NN committed:
refactor: Separate into functions

find_x_vals() and create_plot_data() are extracted from plot_objective()
and made into separate functions

SRFU-NN committed:
feat: Only determining point-bin membership once.

SRFU-NN commented:
To make the dependency plots, the current implementation randomly selects 250 points, and for each pixel changes the value(s) in the plotted directions, calls model.predict on the resulting 250 points, and takes the average.

If we instead take more random points over the entire space, and simply average the ones that happen to be within the value(s) for each pixel, we can dramatically reduce the number of calls to model.predict - We go from n^2 calls to one call, with the same amount of samples predicted per call in the naive approach. With a bit of care as to how we find the ones that are within each pixel, we can make the plotting faster.

That "bit of care" turns out to be quit a bit slower if it is to allow categorical dimensions due to the one hot encoding.

We can also lower the amount of samples to predict by broadening the averaging window. Since the kernel ensures that both the mean of the Gaussian process and the variance are continuous, averaging over, say, the width of three pixels shouldn't introduce too much change. In the 2D plots, this means taking the average of 9 times more points, so the number of samples can be reduced by this factor.

Note that my timing is not as reliable as it ought to be, it depends intimately on what else is running on my computer, so verification would be appreciated.

SRFU-NN commented:
Basic plotting with both continuous and categorical dimensions, existing code, 250 samples per pixel. Takes around 10 seconds to create:
![Figure 1 (3)](https://user-images.githubusercontent.com/104423389/206439456-e7dc9596-d872-441f-8f77-3eaf36bfba2e.png)
Same plot with new code, 10 samples per pixel, averaging over the neighbouring pixels for a total of 90 samples per pixel, takers around 6 seconds to create:
![output](https://user-images.githubusercontent.com/104423389/206439558-c80a9c37-6ede-43b7-babd-91c3645c6133.png)

SRFU-NN commented:
The codes is faster if no categorical dimension is included

With existing code:
![Figure 3](https://user-images.githubusercontent.com/104423389/206448516-2ef47aa0-8dce-4871-a6f0-61bcfc5c112c.png)

With new code, 3.7 seconds: 
![output](https://user-images.githubusercontent.com/104423389/206448584-c5293d14-5821-499e-8312-188fd550af0a.png)

SRFU-NN commented:
For plots with a set X point (not partial dependence), the noise level drops, as we are taking the value in a point. So the amount of samples can be dropped a lot, reducing the time for the previous example to around 1.5 seconds. But I assume a similar thing can be done with the existing code.

SRFU-NN commented:
Currently, the default oversampling (the average number of samples per pixel) is set to 10 for partial dependence, and 2 for a set value (this is mostly due to how the dimensions are sampled, which should probably be fixed instead of forcing oversampling to be over 1). value of 10 is just at the boarder of what is an acceptable level of noise

There is a tradeoff between time and noise level with regard to oversampling. Higher oversampling leads to less noise (theoretically by the square root of the oversampling), but takes longer to make (presumably linear with a constant added). The oversampling = 30 plot takes around twice as long as the oversampling = 10 plot

Oversampling = 30 plot:
![output](https://user-images.githubusercontent.com/104423389/206452302-a8342c3d-1698-4623-82bf-af155cbad815.png)

Oversampling = 10 plot: 
![output](https://user-images.githubusercontent.com/104423389/206452416-993b2a56-3f0f-4998-b0f7-665b41c2833e.png)

SRFU-NN commented:
A main question is: Do we want to keep the partial dependence plotting (taking the average over all of the non-plotted dimensions), or other kinds of "summarising" plots (e.g. minimum over the non-plotted dimensions)? If not, this pull request is probably not worth it.

dk-teknologisk-mon commented:
I tried to take a first look at the files, but I think you may have to re-do the pull request (sorry), since something has gone wrong with  `plots.py`. It registers as having been completely replaced and as a result I cannot really see where your changes have been made. Before making a new pull request, you should also fetch the latest version of develop, since it contains a lot of changes to `plots.py` compared to the version you are trying to merge. If we just accept this request, those changes would get overwritten again.

SRFU-NN commented:
Dammit!

We also need to discuss whether we want this pull request – As I wrote the comments, I figured out that the plotting for a specified X can be done a lot simpler by simply taking one point per pixel, so it is solely a way to make partial dependency plot faster, and I am not convinced we want to keep supporting that.

Should we have aa talk later this week?

Best regards
Søren Furbo
Automation Scientist
7660 Automation and Process Optimisation
+45 3079 0427 (direct) ǀ  ***@***.******@***.***>


From: Morten Bormann Nielsen ***@***.***>
Sent: 12. december 2022 09:20
To: novonordisk-research/ProcessOptimizer ***@***.***>
Cc: SRFU (Søren Furbo) ***@***.***>; Assign ***@***.***>
Subject: Re: [novonordisk-research/ProcessOptimizer] Fix#103 (PR #133)

Caution: This email originated from outside of the Novo Nordisk email system. Use caution when clicking on links or opening attachments.


I tried to take a first look at the files, but I think you may have to re-do the pull request (sorry), since something has gone wrong with plots.py. It registers as having been completely replaced and as a result I cannot really see where your changes have been made. Before making a new pull request, you should also fetch the latest version of develop, since it contains a lot of changes to plots.py compared to the version you are trying to merge. If we just accept this request, those changes would get overwritten again.

—
Reply to this email directly, view it on GitHub<https://github.com/novonordisk-research/ProcessOptimizer/pull/133#issuecomment-1346068929>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AY4V7XM7CDKHZCJE4DZ4VCLWM3NZPANCNFSM6AAAAAASQYYAO4>.
You are receiving this because you were assigned.Message ID: ***@***.******@***.***>>

SRFU-NN committed:
Merge remote-tracking branch 'origin/develop' into Fix#103

SRFU-NN commented:
I have done a manual merge of the two branches and pushed that, but GitHub till sees it as a completely new file, even when I make a new pull request.

I can try and make a new branch based on develop, and copy the merged plots.py into it, but I am not optimistic. There might simply be too many differences for Github to understand what is going on.

Git in my VS Code can handle it, so perhaps we can use a different tool?

Best regards
Søren Furbo
Automation Scientist
7660 Automation and Process Optimisation
+45 3079 0427 (direct) ǀ  ***@***.******@***.***>


From: Morten Bormann Nielsen ***@***.***>
Sent: 12. december 2022 09:20
To: novonordisk-research/ProcessOptimizer ***@***.***>
Cc: SRFU (Søren Furbo) ***@***.***>; Assign ***@***.***>
Subject: Re: [novonordisk-research/ProcessOptimizer] Fix#103 (PR #133)

Caution: This email originated from outside of the Novo Nordisk email system. Use caution when clicking on links or opening attachments.


I tried to take a first look at the files, but I think you may have to re-do the pull request (sorry), since something has gone wrong with plots.py. It registers as having been completely replaced and as a result I cannot really see where your changes have been made. Before making a new pull request, you should also fetch the latest version of develop, since it contains a lot of changes to plots.py compared to the version you are trying to merge. If we just accept this request, those changes would get overwritten again.

—
Reply to this email directly, view it on GitHub<https://github.com/novonordisk-research/ProcessOptimizer/pull/133#issuecomment-1346068929>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AY4V7XM7CDKHZCJE4DZ4VCLWM3NZPANCNFSM6AAAAAASQYYAO4>.
You are receiving this because you were assigned.Message ID: ***@***.******@***.***>>

SRFU-NN committed:
fix: Readded fancy legend

SRFU-NN committed:
fix: evenly_sample border list for transformed dim

The borde list of evenly_sampl of dimensions assumed that the dimension
was normalised. Now it works with any transformation.

Inclduing fixing of tests, and testing or the different options for
transformations.

SRFU-NN committed:
test: NOW tests are updated

SRFU-NN committed:
fix: locking sklearn version to and locate error

SRFU-NN committed:
fix: Replaced mse with friedman_mse

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -313,6 +317,44 @@ def lhs_arange(self, n):
+            The number of points to sample.
+
+        * `border_width` [int]
+            How wide to make the borders. If None, no borders are returned.
```
I don't understand why this number should be an int? Are the borders not generally a fraction of the dimension length? If not, then how are they defined? What happens when a dimension is, say, [0.0,  1.0]?

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -313,6 +317,44 @@ def lhs_arange(self, n):
+            The transformed values of `xi`, for feeding to a model.
+
+        * `border_list`: [list of tuples]:
+            A list of where 
```
The explanation for border_list is truncated here. I guess it should be made clear that this list of touples is not always returned, depending on border_width?

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -313,6 +317,44 @@ def lhs_arange(self, n):
+            return xi, xi_transformed
+        border_list = []
+        transformed_bounds = self.transform(self.bounds)
+        border_distance = float(
```
From this point and the next lines down, could use adherence to PEP standard (last ")" that closes float onto its own line, spaces around + and -).

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -420,14 +462,50 @@ def lhs_arange(self, n):
+            The number of points to sample from `dim`.
+
+        * `border_width` [int]
+            How wide to make the borders. If None, no borders are returned.
```
Again, is it intentional that this input is an int? Is the border simply in the relevant dimension scale?

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -420,14 +462,50 @@ def lhs_arange(self, n):
+        border_list = []
+        transformed_bounds = self.transform(self.bounds)
+        border_distance = float(
+            np.diff(transformed_bounds)/(num_points-1)*border_width)
```
Same PEP comment as above, the closing paranthesis of the float argument should be moved down on its own line, and inside the for-loop we should add spaces around the operators + and -.

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -566,6 +644,37 @@ def lhs_arange(self, n):
+        * `num_points` [int]
+            The number of points to sample.
+
+        * `border_width` [int]
```
How does an integer border width make sense for categories?

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -566,6 +644,37 @@ def lhs_arange(self, n):
+        if border_width is None:
+            return xi, xi_transformed
+        # For categorial dimensions, it doesn't make
+        border_list = [(x, x) for x in xi_transformed]
```
Why is the comment split by this line?

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_space.py:
```diff
@@ -697,3 +697,48 @@ def test_lhs():
+    dim = dimension_type(-10, 10, transform=transform)
+    xi, xi_transformed, border_list = dim.evenly_sample(
+        num_points=num_points,
+        border_width=border_width)
```
Closing ) onto its own line.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_space.py:
```diff
@@ -697,3 +697,48 @@ def test_lhs():
+    assert np.allclose(upper_bound_borders, xi_transformed[-1])
+    # Testing that lower borders are coincide with the correct xi
+    lower_borders = [
+        border_list[i+border_width][0]for i in range(num_points-border_width)
```
Missing space in front of the "for"

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_space.py:
```diff
@@ -697,3 +697,48 @@ def test_lhs():
+        lower_borders, xi_transformed[0:num_points-border_width])
+    # Testing that upper borders are coincide with the correct xi
+    upper_borders = [
+        border_list[i][1]for i in range(num_points-border_width)
```
Also missing space in front of the "for" here.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_space.py:
```diff
@@ -697,3 +697,48 @@ def test_lhs():
+    dim = Categorical(["cat", "mouse", "dog"])
+    xi, xi_transformed, border_list = dim.evenly_sample(
+        num_points=num_points,
+        border_width=border_width)
```
Closing ) onto its own line.

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+                            lw=1,
+                        )
+                        ax[0][-1].legend(
+                            handles=[legend_data_point,
```
Do we really need the strange newline construction here? If you want the contents of the `handles` list to appear legible we should at least split them across a new line for each entry, instead of just having a single new line after the first argument. The way this is now structured makes it harder to decode the individual arguments for the legend call.

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+                        else:
+                            ci_label = "95 % credibility interval"
+                        ax[0][-1].legend(
+                            handles=[legend_data_point,
```
Do we really need the strange newline construction here? If you want the contents of the `handles` list to appear legible we should at least split them across a new line for each entry, instead of just having a single new line after the first argument. The way this is now structured makes it harder to decode the individual arguments for the legend call.

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+                        vmin=limits["z_min"], vmax=limits["z_max"]
+                    )
+                    cb = ax[0][-1].figure.colorbar(
+                        mpl.cm.ScalarMappable(
```
If you want to break the single line formatting here, we should do so for each argument to ScalarMappable and ensure the closing parenthesis ends on its own line.

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+        Number of levels to draw on the contour plot, passed directly to
+        `plt.contour()`.
+
+    * `oversampling` [int, default=10]
```
The default set in the function definition is None, not 10. Either change it here or in the function default.

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+            "result" - Use best observed parameters
+            "expected_minimum" - Parameters that gives the best minimum. Calculated
+                using scipy's minimize method. This method currently does not work with
+                categorical values.
```
*with purely categorical values. It works fine if there's just one.

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+    * `title` [str, default=None]
+        String to use as title of the figure.
+
+    * `show_confidence` [bool, default=False]
```
I still think we should change this default to True.

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+        * "interpolation" [string, default=""]: The interplotation visaulization to use for the 2D
+            plots. For possible values, see
+            https://matplotlib.org/stable/gallery/images_contours_and_fields/interpolation_methods.html
+            If empty string, a contour pot is shown. Note that contour plots are not
```
"plot" not "pot". It should be made clear that contour plots are not compatible with showing uncertainty _in the 2D plots_.

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+            compatible with showing uncertainty.
+        * "uncertain_color"  [any matplotlib color, default=[0,0,0]]: The colour to use
+            for "very" uncertain values in 2D plots.
+        * "colormap" [string, default="viridis_r"]: The colourmap to use. For possible
```
The colormap to use in the 2D plots.*

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+def plot_objective(
+    result,
+    levels=10,
+    oversampling=None,
```
Do we really need to bloat the function call with user input on oversampling? I think we should develop good settings that simply live inside the function instead of allowing user input on this. We don't provide handles for changing aestetic aspects of the plots (color of points etc.) for the same reason.

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+        "uncertain_color": [0, 0, 0],
+        "colormap": "viridis_r",
+        "normalize_uncertainty": (
+            lambda x, global_min, global_max: (
```
The single hanging ( at the end of the line looks really weird when reading the code here.

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+                # Use standard of 20 random starting points
+                num_random_starts = 20
+            x_vals, _ = expected_minimum(
+                result, n_random_starts=num_random_starts)
```
Inputs need correct formatting here. What is the reason that `random_state=None` was removed?

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+                # If a value for expected_minimum_samples has been parsed
+                num_random_starts = expected_minimum_samples
+            else:
+                # Use standard of 10^n_parameters. Note this becomes very slow
```
The comment and the code disagree here. I'm fine with hardcoding 1E5 as a value.

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+    plot_options["zscale"] = zscale
+    plot_options["levels"] = levels
+
+    x_vals = find_x_vals(pars, result, expected_minimum_samples)
```
Maybe add a comment here to the effect of "Identify the location of the point to highlight in the plots".

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+    return plots_data, limits
+
+
+def create_sample_list(space, graph_width, oversampling):
```
This function needs a lot more comments and a lot less list comprehension for-loops to be readable. I simply cannot follow what is going on in here.

I also think you need to think a little more about how you name variables in here. I mean what am I to infer from the name `included_list_list`?

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+    sampling_dimensions = [
+        dim.evenly_sample(graph_width*oversampling)[1]
+        for dim in space.dimensions]
+    sample_list = [
```
This is also a monster of deep list comprehension where the formatting of the call makes it really hard for me to understand what is going on.

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+    average_value_list = []
+    average_stddev_list = []
+    for included_samples in included_list_list:
+        # Add a colorbar for the 2D plot value scale
```
Does this comment belong here?

dk-teknologisk-mon commented on ProcessOptimizer/plots.py:
```diff
@@ -1,2382 +1,2493 @@
+    return x_vals
+
+
+def create_plot_data(oversampling, graph_width, space, model, usepartialdependence, show_confidence, x_vals):
```
On the high level: It's fine that you have abstracted out the internal parts that gather the data for the plot, but in doing so you have replaced the use of the comprehensively documented `dependence` function for calls of `generate_random_data`, `find_average_1D` and `find_average_2D`, which have very limited documentation as to how they work and are very hard to read. If the problem is that dependence is slow, surely it would be better to fix its inner workings rather than leaving it as is and building a new function that does its work for it? I know this isn't exactly "fun" feedback to get now you have done a lot of work to speed things up, but it bloats the code base when we just create new functionality instead of fixing the stuff that we think runs too slowly.
